src/main.rs

```diff
@@ -85,33 +58,80 @@ async fn run_single_slot(
     Ok(())
 }
 
+async fn run_oracle_node(
+    price_provider: Arc<impl PriceProvider>,
+    message_generator: MessageGenerator,
+    message_broadcaster: Arc<impl MessageBroadcaster>,
+    slot_provider: Arc<SystemClockSlotProvider>,
+    rx: mpsc::Receiver<()>,
+) -> Result<()> {
+    ReceiverStream::new(rx)
+        .for_each_concurrent(MAX_CONCURRENT_SLOTS, |_| async {
+            if let Some(slot) = slot_provider.get_last_slot().await {
+                log::debug!(
+                    "processing slot with number: {}, slot start: {}, processing started at: {}",
+                    slot.number,
+                    slot.to_date_time(),
+                    Utc::now()
+                );
+                timeout(
+                    tokio::time::Duration::from_secs(RUN_SLOT_LIMIT_SECS),
+                    run_single_slot(
+                        price_provider.clone(),
+                        message_generator.clone(),
+                        message_broadcaster.clone(),
+                        slot,
+                    ),
+                )
+                .await
+                .unwrap_or_else(|_| {
+                    log::error!(
+                        "Hit {}s timeout for slot: {}",
+                        RUN_SLOT_LIMIT_SECS,
+                        slot.number
+                    );
+                    Ok(())
+                })
+                .unwrap_or_else(|e| {
+                    log::error!("Error when running for slot: {} - {:?}", slot.number, e);
+                });
+            }
+        })
+        .await;
+    Ok(())
+}
+
 #[tokio::main]
 async fn main() -> Result<()> {
     env_logger::init();
 
-    let mut gofer_cmd = std::env::var("GOFER_CMD")?; // Assumes that GOFER_CMD env variable is set
-                                                     // to the gofer binary
+    // Assumes that GOFER_CMD env variable is set to the gofer binary
```

Comment:
Moved it mostly for consistency but in general I much prefer comments to go above lines not next to.
* plays nicer with virtual text return types written next to lines
* horizontal space is scarce, vertical space is abundant
* rustfmt apparently uses whitespace aligning on comments, neat but also wtf, breaks linewise scanning for your eyes

Reply:
Agree, lets go with comments above lines.